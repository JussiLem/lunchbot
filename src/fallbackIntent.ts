import { aws_lex as lex } from 'aws-cdk-lib'

export const fallbackIntent: lex.CfnBot.IntentProperty = {
  description: 'Default intent when no other intent matches',
  initialResponseSetting: {
    nextStep: {
      dialogAction: {
        type: 'InvokeDialogCodeHook',
      },
    },
    codeHook: {
      isActive: true,
      enableCodeHookInvocation: true,
      postCodeHookSpecification: {
        successNextStep: {
          dialogAction: {
            type: 'EndConversation',
          },
        },
        failureNextStep: {
          dialogAction: {
            type: 'EndConversation',
          },
        },
      },
    },
  },
  name: 'FallbackIntent',
  parentIntentSignature: 'AMAZON.FallbackIntent',
}
