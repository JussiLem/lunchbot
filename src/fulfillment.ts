// eslint-disable-next-line import/no-extraneous-dependencies
import { LexV2Handler, LexV2Intent, LexV2Result } from 'aws-lambda'
import { logger } from './common/powertools'
// eslint-disable-next-line import/no-extraneous-dependencies
import { LexV2ScalarSlotValue, LexV2Slot } from 'aws-lambda/trigger/lex-v2'

interface SuggestLunchSlots {
  CuisineType?: LexV2ScalarSlotValue | null
  DietaryRestrictions?: LexV2ScalarSlotValue | null
  Budget?: LexV2ScalarSlotValue | null
}

/**
 * Ensure session state dialog action is set to Delegate, pass everything as-is
 */
const delegate = (
  sessionAttributes: Record<string, string> | undefined,
  intent: LexV2Intent,
) =>
  ({
    sessionState: {
      intent,
      sessionAttributes,
      dialogAction: {
        type: 'Delegate',
      },
    },
  }) as LexV2Result

/**
 * Type guard to ensure slot contains the expected structure
 */
const isSlotValue = (slot: LexV2Slot): slot is LexV2ScalarSlotValue => {
  return (
    slot &&
    typeof slot === 'object' &&
    'value' in slot &&
    'originalValue' in slot.value &&
    Array.isArray(slot.value.resolvedValues) &&
    'interpretedValue' in slot.value
  )
}

export const handler: LexV2Handler = async (event) => {
  logger.info('Received event', {
    event,
  })
  const { sessionState } = event
  const { sessionAttributes, intent } = sessionState
  const { name: intentName, slots } = intent
  logger.info('Current intent and slots:', {
    intentName,
    slots,
  })
  if (slots) {
    const suggestLunchSlots = slots as SuggestLunchSlots

    if (
      suggestLunchSlots.CuisineType &&
      isSlotValue(suggestLunchSlots.CuisineType)
    ) {
      const cuisineType = suggestLunchSlots.CuisineType.value.interpretedValue
      logger.info('Cuisine type detected:', { cuisineType })

      // Additional logic based on CuisineType can be added here
      return delegate(sessionAttributes, intent)
    }
  }
  return delegate(sessionAttributes, intent)
}
