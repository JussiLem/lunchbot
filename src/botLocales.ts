import { aws_lex as lex } from 'aws-cdk-lib'
import { fallbackIntent } from './fallbackIntent'

const intents: lex.CfnBot.IntentProperty[] = [
  fallbackIntent,
  {
    dialogCodeHook: {
      enabled: true,
    },
    name: 'SuggestLunch',
    description: 'Intent to suggest lunch',
    sampleUtterances: [
      { utterance: 'What should I have for lunch?' },
      { utterance: 'Give me lunch suggestions' },
      { utterance: 'Suggest a lunch option' },
      { utterance: "What's for lunch?" },
      { utterance: 'Hungry' },
    ],
    initialResponseSetting: {
      nextStep: {
        dialogAction: {
          type: 'ElicitSlot',
          slotToElicit: 'OfficeLocation',
        },
      },
    },
    slots: [
      {
        slotTypeName: 'OfficeLocations',
        name: 'OfficeLocation',
        valueElicitationSetting: {
          slotConstraint: 'Required',
          promptSpecification: {
            maxRetries: 2,
            messageGroupsList: [
              {
                message: {
                  imageResponseCard: {
                    title: 'Select your office location',
                    imageUrl:
                      'https://www.sttinfo.fi/data/images/public/69817246/70096437/ff783596-4db9-40be-bd62-d1ab7b1d75c1-w_960.jpg',
                    buttons: [
                      {
                        text: 'Kamppi',
                        value: 'Kamppi',
                      },
                      {
                        text: 'Hakaniemi',
                        value: 'Hakaniemi',
                      },
                    ],
                  },
                },
              },
            ],
          },
        },
      },
      {
        slotTypeName: 'CuisineType',
        name: 'CuisineType',
        valueElicitationSetting: {
          slotConstraint: 'Required',
          promptSpecification: {
            maxRetries: 2,
            messageGroupsList: [
              {
                message: {
                  plainTextMessage: {
                    value: 'What type of cuisine are you interested in?',
                  },
                },
              },
            ],
          },
        },
      },
      {
        slotTypeName: 'Restaurants',
        name: 'Restaurants',
        valueElicitationSetting: {
          slotConstraint: 'Required',
          promptSpecification: {
            maxRetries: 2,
            messageGroupsList: [
              {
                message: {
                  plainTextMessage: {
                    value: 'Pick the restaurant',
                  },
                },
              },
            ],
          },
        },
      },
      {
        slotTypeName: 'DietaryRestrictions',
        name: 'DietaryRestrictions',
        valueElicitationSetting: {
          slotConstraint: 'Optional',
          promptSpecification: {
            maxRetries: 2,
            messageGroupsList: [
              {
                message: {
                  plainTextMessage: {
                    value: 'Do you have any dietary restrictions?',
                  },
                },
              },
            ],
          },
        },
      },
      {
        slotTypeName: 'AMAZON.Number',
        name: 'Budget',
        valueElicitationSetting: {
          slotConstraint: 'Optional',
          promptSpecification: {
            maxRetries: 2,
            messageGroupsList: [
              {
                message: {
                  plainTextMessage: { value: 'What is your budget for lunch?' },
                },
              },
            ],
          },
        },
      },
    ],
    slotPriorities: [
      {
        priority: 1,
        slotName: 'OfficeLocation',
      },
      {
        priority: 2,
        slotName: 'CuisineType',
      },
      {
        priority: 3,
        slotName: 'Restaurants',
      },
      {
        priority: 4,
        slotName: 'DietaryRestrictions',
      },
      {
        priority: 5,
        slotName: 'Budget',
      },
    ],
  },
]
export const botLocales: lex.CfnBot.BotLocaleProperty[] = [
  {
    localeId: 'en_US',
    description: 'LunchBot locale',
    nluConfidenceThreshold: 0.4,
    intents,
    slotTypes: [
      {
        name: 'OfficeLocations',
        description: 'Office location',
        valueSelectionSetting: {
          resolutionStrategy: 'ORIGINAL_VALUE',
        },
        slotTypeValues: [
          { sampleValue: { value: 'Kamppi' } },
          { sampleValue: { value: 'Hakaniemi' } },
        ],
      },
      {
        name: 'DietaryRestrictions',
        description: 'Common dietary restrictions',
        valueSelectionSetting: {
          resolutionStrategy: 'ORIGINAL_VALUE',
        },
        slotTypeValues: [
          { sampleValue: { value: 'Vegetarian' } },
          { sampleValue: { value: 'Vegan' } },
        ],
      },
      {
        name: 'Restaurants',
        description: 'Possible lunch restaurants',
        valueSelectionSetting: {
          resolutionStrategy: 'ORIGINAL_VALUE',
        },
        slotTypeValues: [
          { sampleValue: { value: 'Thai Restaurant' } },
          { sampleValue: { value: 'Chinese Restaurant' } },
          { sampleValue: { value: 'Buffet Restaurant' } },
        ],
      },
      {
        valueSelectionSetting: {
          resolutionStrategy: 'ORIGINAL_VALUE',
        },
        description: 'Types of cuisine',
        name: 'CuisineType',
        slotTypeValues: [
          {
            sampleValue: {
              value: 'Italian',
            },
          },
          {
            sampleValue: {
              value: 'Chinese',
            },
          },
          {
            sampleValue: {
              value: 'Mexican',
            },
          },
          {
            sampleValue: {
              value: 'Indian',
            },
          },
          {
            sampleValue: {
              value: 'American',
            },
          },
        ],
      },
    ],
  },
]
