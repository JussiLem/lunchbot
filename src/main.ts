import {
  App,
  aws_iam as iam,
  aws_lambda as lambda,
  aws_lambda_nodejs as nodejs,
  aws_lex as lex,
  Stack,
  StackProps,
} from 'aws-cdk-lib'
import { Construct } from 'constructs'
import { botLocales } from './botLocales'

export class LunchBotStack extends Stack {
  constructor(
    scope: Readonly<Construct>,
    id: string,
    props: Readonly<StackProps> = {},
  ) {
    super(scope, id, props)
    const lexBotRole = new iam.Role(this, 'LexBotRole', {
      assumedBy: new iam.ServicePrincipal('lexv2.amazonaws.com'),
      inlinePolicies: {
        ['LexRuntimeRolePolicy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              resources: ['*'],
              actions: ['comprehend:DetectSentiment'],
            }),
          ],
        }),
      },
    })
    lexBotRole.addToPolicy(
      new iam.PolicyStatement({
        actions: [
          'lex:*',
          'lex:CreateBotLocale',
          'lex:CreateBot',
          'lex:StartImport',
          'lex:DescribeBot',
          'lex:DescribeBotLocale',
          'lex:DescribeBotVersion',
          'lex:ListBots',
          'lex:ListBotAliases',
          'lex:ListBotLocales',
          'lex:ListBotVersions',
          'lex:PutBotAlias',
          'lex:PutBotChannelAssociation',
        ],
        resources: ['*'],
      }),
    )
    const fulfillmentLambda = new nodejs.NodejsFunction(
      this,
      'fulfillmentLambda',
      {
        runtime: lambda.Runtime.NODEJS_20_X,
        environment: {
          SERVICE_NAME: 'lunchbot',
          POWERTOOLS_LOG_LEVEL: 'DEBUG',
        },
        entry: 'src/fulfillment.ts',
      },
    )
    const lunchBot = new lex.CfnBot(this, 'LunchBot', {
      dataPrivacy: { ChildDirected: true },
      idleSessionTtlInSeconds: 300,
      name: 'LunchBot',
      roleArn: lexBotRole.roleArn,
      autoBuildBotLocales: true,
      botLocales,
      testBotAliasSettings: {
        botAliasLocaleSettings: [
          {
            localeId: 'en_US',
            botAliasLocaleSetting: {
              enabled: true,
              codeHookSpecification: {
                lambdaCodeHook: {
                  lambdaArn: fulfillmentLambda.functionArn,
                  codeHookInterfaceVersion: '1.0',
                },
              },
            },
          },
        ],
      },
    })

    const botVersion = new lex.CfnBotVersion(this, 'LunchBotVersion', {
      botId: lunchBot.ref,
      botVersionLocaleSpecification: [
        {
          botVersionLocaleDetails: {
            sourceBotVersion: 'DRAFT',
          },
          localeId: 'en_US',
        },
      ],
    })
    new lex.CfnBotAlias(this, 'devBotAlias', {
      botAliasName: 'dev',
      botId: lunchBot.ref,
      botAliasLocaleSettings: [
        {
          botAliasLocaleSetting: {
            enabled: true,
            codeHookSpecification: {
              lambdaCodeHook: {
                lambdaArn: fulfillmentLambda.functionArn,
                codeHookInterfaceVersion: '1.0',
              },
            },
          },
          localeId: 'en_US',
        },
      ],
      botVersion: botVersion.attrBotVersion,
      sentimentAnalysisSettings: { DetectSentiment: true },
    })
    fulfillmentLambda.addPermission('Lex Invocation', {
      principal: new iam.ServicePrincipal('lexv2.amazonaws.com'),
      sourceArn: `arn:aws:lex:${Stack.of(this).region}:${Stack.of(this).account}:bot-alias/${lunchBot.attrId}/*`,
    })
    // const eventLambda = new nodejs.NodejsFunction(this, 'events', {
    //   runtime: lambda.Runtime.NODEJS_20_X,
    //   environment: {
    //     BOT_ID: botAlias.botId,
    //     BOT_ALIAS_ID: botAlias.attrBotAliasId,
    //     POWERTOOLS_LOG_LEVEL: 'DEBUG',
    //   },
    //   entry: path.join('src/listen-slack-events/listen-slack-events.api.ts'),
    // })
    // Attach Lex permissions to the Lambda role
    // eventLambda.addToRolePolicy(
    //   new iam.PolicyStatement({
    //     actions: [
    //       'lex:StartConversation',
    //       // Add other Lex permissions if needed
    //     ],
    //     resources: [
    //       `arn:aws:lex:${Stack.of(this).region}:${Stack.of(this).account}:bot-alias/${lunchBot.attrId}/*`,
    //     ],
    //   }),
    // )
  }
}

const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
}

const app = new App()

new LunchBotStack(app, 'lunchbot-dev', { env: devEnv })

app.synth()
