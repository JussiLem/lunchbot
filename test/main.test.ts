import { App } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { LunchBotStack } from '../src/main'

// Expected patterns
const assetMatch = /[\da-f]{64}\.zip/
const bucketMatch = new RegExp(`cdk-[0-9a-z]{9}-assets`)

expect.addSnapshotSerializer({
  test: (val) =>
    typeof val === 'string' &&
    (val.match(bucketMatch) != null || val.match(assetMatch) != null),
  print: (val) => {
    // Substitute both the bucket part and the asset zip part
    // eslint-disable-next-line functional/no-let
    let sval = `${val}`
    sval = sval.replace(bucketMatch, '[ASSET BUCKET]')
    sval = sval.replace(assetMatch, '[ASSET ZIP]')
    return `"${sval}"`
  },
})
test('Snapshot', () => {
  const app = new App()
  const stack = new LunchBotStack(app, 'test')

  const template = Template.fromStack(stack)
  expect(template.toJSON()).toMatchSnapshot()
})
