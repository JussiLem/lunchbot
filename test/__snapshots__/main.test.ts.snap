// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LexBotRole0FBFF464": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lexv2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "comprehend:DetectSentiment",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LexRuntimeRolePolicy",
          },
        ],
        "Tags": [
          {
            "Key": "AppManagerCFNStackKey",
            "Value": "test",
          },
          {
            "Key": "Environment",
            "Value": "dev",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LexBotRoleDefaultPolicyE0530F06": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "lex:*",
                "lex:CreateBotLocale",
                "lex:CreateBot",
                "lex:StartImport",
                "lex:DescribeBot",
                "lex:DescribeBotLocale",
                "lex:DescribeBotVersion",
                "lex:ListBots",
                "lex:ListBotAliases",
                "lex:ListBotLocales",
                "lex:ListBotVersions",
                "lex:PutBotAlias",
                "lex:PutBotChannelAssociation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LexBotRoleDefaultPolicyE0530F06",
        "Roles": [
          {
            "Ref": "LexBotRole0FBFF464",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LunchBot": {
      "Properties": {
        "AutoBuildBotLocales": true,
        "BotLocales": [
          {
            "Description": "LunchBot locale",
            "Intents": [
              {
                "Description": "Default intent when no other intent matches",
                "InitialResponseSetting": {
                  "CodeHook": {
                    "EnableCodeHookInvocation": true,
                    "IsActive": true,
                    "PostCodeHookSpecification": {
                      "FailureNextStep": {
                        "DialogAction": {
                          "Type": "EndConversation",
                        },
                      },
                      "SuccessNextStep": {
                        "DialogAction": {
                          "Type": "EndConversation",
                        },
                      },
                    },
                  },
                  "NextStep": {
                    "DialogAction": {
                      "Type": "InvokeDialogCodeHook",
                    },
                  },
                },
                "Name": "FallbackIntent",
                "ParentIntentSignature": "AMAZON.FallbackIntent",
              },
              {
                "Description": "Intent to suggest lunch",
                "DialogCodeHook": {
                  "Enabled": true,
                },
                "InitialResponseSetting": {
                  "NextStep": {
                    "DialogAction": {
                      "SlotToElicit": "OfficeLocation",
                      "Type": "ElicitSlot",
                    },
                  },
                },
                "Name": "SuggestLunch",
                "SampleUtterances": [
                  {
                    "Utterance": "What should I have for lunch?",
                  },
                  {
                    "Utterance": "Give me lunch suggestions",
                  },
                  {
                    "Utterance": "Suggest a lunch option",
                  },
                  {
                    "Utterance": "What's for lunch?",
                  },
                ],
                "SlotPriorities": [
                  {
                    "Priority": 1,
                    "SlotName": "OfficeLocation",
                  },
                  {
                    "Priority": 2,
                    "SlotName": "CuisineType",
                  },
                  {
                    "Priority": 3,
                    "SlotName": "DietaryRestrictions",
                  },
                  {
                    "Priority": 4,
                    "SlotName": "Budget",
                  },
                ],
                "Slots": [
                  {
                    "Name": "OfficeLocation",
                    "SlotTypeName": "OfficeLocations",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "MaxRetries": 2,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "ImageResponseCard": {
                                "Buttons": [
                                  {
                                    "Text": "Kamppi",
                                    "Value": "Kamppi",
                                  },
                                  {
                                    "Text": "Hakaniemi",
                                    "Value": "Hakaniemi",
                                  },
                                ],
                                "ImageUrl": "https://www.sttinfo.fi/data/images/public/69817246/70096437/ff783596-4db9-40be-bd62-d1ab7b1d75c1-w_960.jpg",
                                "Title": "Select your office location",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                  {
                    "Name": "CuisineType",
                    "SlotTypeName": "CuisineType",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "MaxRetries": 2,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What type of cuisine are you interested in?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                  {
                    "Name": "DietaryRestrictions",
                    "SlotTypeName": "DietaryRestrictions",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "MaxRetries": 2,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "Do you have any dietary restrictions?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Optional",
                    },
                  },
                  {
                    "Name": "Budget",
                    "SlotTypeName": "AMAZON.Number",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "MaxRetries": 2,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What is your budget for lunch?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Optional",
                    },
                  },
                ],
              },
            ],
            "LocaleId": "en_US",
            "NluConfidenceThreshold": 0.4,
            "SlotTypes": [
              {
                "Description": "Office location",
                "Name": "OfficeLocations",
                "SlotTypeValues": [
                  {
                    "SampleValue": {
                      "Value": "Kamppi",
                    },
                  },
                  {
                    "SampleValue": {
                      "Value": "Hakaniemi",
                    },
                  },
                ],
                "ValueSelectionSetting": {
                  "ResolutionStrategy": "ORIGINAL_VALUE",
                },
              },
              {
                "Description": "Common dietary restrictions",
                "Name": "DietaryRestrictions",
                "SlotTypeValues": [
                  {
                    "SampleValue": {
                      "Value": "Vegetarian",
                    },
                  },
                  {
                    "SampleValue": {
                      "Value": "Vegan",
                    },
                  },
                ],
                "ValueSelectionSetting": {
                  "ResolutionStrategy": "ORIGINAL_VALUE",
                },
              },
              {
                "Description": "Types of cuisine",
                "Name": "CuisineType",
                "SlotTypeValues": [
                  {
                    "SampleValue": {
                      "Value": "Italian",
                    },
                  },
                  {
                    "SampleValue": {
                      "Value": "Chinese",
                    },
                  },
                  {
                    "SampleValue": {
                      "Value": "Mexican",
                    },
                  },
                  {
                    "SampleValue": {
                      "Value": "Indian",
                    },
                  },
                  {
                    "SampleValue": {
                      "Value": "American",
                    },
                  },
                ],
                "ValueSelectionSetting": {
                  "ResolutionStrategy": "ORIGINAL_VALUE",
                },
              },
            ],
          },
        ],
        "BotTags": [
          {
            "Key": "AppManagerCFNStackKey",
            "Value": "test",
          },
          {
            "Key": "Environment",
            "Value": "dev",
          },
        ],
        "DataPrivacy": {
          "ChildDirected": true,
        },
        "IdleSessionTTLInSeconds": 300,
        "Name": "LunchBot",
        "RoleArn": {
          "Fn::GetAtt": [
            "LexBotRole0FBFF464",
            "Arn",
          ],
        },
        "TestBotAliasSettings": {
          "BotAliasLocaleSettings": [
            {
              "BotAliasLocaleSetting": {
                "CodeHookSpecification": {
                  "LambdaCodeHook": {
                    "CodeHookInterfaceVersion": "1.0",
                    "LambdaArn": {
                      "Fn::GetAtt": [
                        "LunchfulfillmentLambdaB53264CB",
                        "Arn",
                      ],
                    },
                  },
                },
                "Enabled": true,
              },
              "LocaleId": "en_US",
            },
          ],
        },
      },
      "Type": "AWS::Lex::Bot",
    },
    "LunchBotVersion": {
      "Properties": {
        "BotId": {
          "Ref": "LunchBot",
        },
        "BotVersionLocaleSpecification": [
          {
            "BotVersionLocaleDetails": {
              "SourceBotVersion": "DRAFT",
            },
            "LocaleId": "en_US",
          },
        ],
      },
      "Type": "AWS::Lex::BotVersion",
    },
    "LunchStateTable49BC594E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          {
            "AttributeName": "slot",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "slot",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Delete",
    },
    "LunchTableFC372FD3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "officeLocation",
            "AttributeType": "S",
          },
          {
            "AttributeName": "cuisineType",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "GSI_OfficeLocation_CuisineType",
            "KeySchema": [
              {
                "AttributeName": "officeLocation",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "cuisineType",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "officeLocation",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "cuisineType",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "GlobalSecondaryIndexes": [
              {
                "IndexName": "GSI_OfficeLocation_CuisineType",
              },
            ],
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Delete",
    },
    "LunchfulfillmentLambdaB53264CB": {
      "DependsOn": [
        "LunchfulfillmentLambdaServiceRoleDefaultPolicyA1B832DE",
        "LunchfulfillmentLambdaServiceRole5FECAB66",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "[ASSET BUCKET]-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[ASSET ZIP]",
        },
        "Environment": {
          "Variables": {
            "LUNCH_TABLE": {
              "Ref": "LunchTableFC372FD3",
            },
            "POWERTOOLS_LOG_LEVEL": "DEBUG",
            "SERVICE_NAME": "lunchbot",
            "STATE_TABLE": {
              "Ref": "LunchStateTable49BC594E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LunchfulfillmentLambdaServiceRole5FECAB66",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "AppManagerCFNStackKey",
            "Value": "test",
          },
          {
            "Key": "Environment",
            "Value": "dev",
          },
        ],
      },
      "Type": "AWS::Lambda::Function",
    },
    "LunchfulfillmentLambdaLexInvocationE461C628": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "LunchfulfillmentLambdaB53264CB",
            "Arn",
          ],
        },
        "Principal": "lexv2.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:lex:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":bot-alias/",
              {
                "Fn::GetAtt": [
                  "LunchBot",
                  "Id",
                ],
              },
              "/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LunchfulfillmentLambdaServiceRole5FECAB66": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "AppManagerCFNStackKey",
            "Value": "test",
          },
          {
            "Key": "Environment",
            "Value": "dev",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LunchfulfillmentLambdaServiceRoleDefaultPolicyA1B832DE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "LunchStateTable49BC594E",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "LunchTableFC372FD3",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "LunchTableFC372FD3",
                          "Arn",
                        ],
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LunchfulfillmentLambdaServiceRoleDefaultPolicyA1B832DE",
        "Roles": [
          {
            "Ref": "LunchfulfillmentLambdaServiceRole5FECAB66",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "devBotAlias": {
      "Properties": {
        "BotAliasLocaleSettings": [
          {
            "BotAliasLocaleSetting": {
              "CodeHookSpecification": {
                "LambdaCodeHook": {
                  "CodeHookInterfaceVersion": "1.0",
                  "LambdaArn": {
                    "Fn::GetAtt": [
                      "LunchfulfillmentLambdaB53264CB",
                      "Arn",
                    ],
                  },
                },
              },
              "Enabled": true,
            },
            "LocaleId": "en_US",
          },
        ],
        "BotAliasName": "dev",
        "BotId": {
          "Ref": "LunchBot",
        },
        "BotVersion": {
          "Fn::GetAtt": [
            "LunchBotVersion",
            "BotVersion",
          ],
        },
        "SentimentAnalysisSettings": {
          "DetectSentiment": true,
        },
      },
      "Type": "AWS::Lex::BotAlias",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
